### ISSUE TITLE
Download directory of vector maps has moved

### ISSUE THREAD

**Issue by Olf0** [rinigus/osmscout-server#246]
OSM Scout Server currently fails to download vector maps, as the name of the download directory on http://osm-maps.kaktux.de/ was changed from *mapboxgl-2* to *mapboxgl-3*, hence the directory  *mapboxgl-2* does not exist anymore.

**Comment by rinigus** [comment]
That's normal, every time I update the maps, the names change. This is done to prevent issues induced by mix of maps from different imports. 

Please press "Check for updates" - this should update the names of the folders and map descriptions. You would have to update the maps after that, if they need updated. In general, kaktux is expected to be used iff modrana is down. So, you would probably get redirected to modrana as well.

**Comment by Olf0** [comment]
Oops, reading thoroughly (as in RTFM) helps.
Not my day, today: Erring too often.

**Comment by rinigus** [comment]
:) no problem. btw, module check is implemented and will be bugging you only if the settings are asking for it .

**Comment by Olf0** [comment]
Someting is strange though: 
1. Coming from the profile "libosmscout with Geocoder-NLP", I switched to "Custom" and activated vector maps.  After that OSM Scout Server correctly wanted to download the vector tiles data for the subscribed regions (plus some updates of libosmscout raster tiles).  Thus I started downloading them.
2. But then OSM Scout Server's Map Manager ran into a download loop: It tried to download some (the last three) updates (I guess the raster ones) again and again, stopping after about 15% and jumping to the second one; the same for the second update, but jumping to the first one after ceasing to download.  The last update was never reached.
3. I remembered that you long ago advised to unsubscribe and re-subscribe the subscribed regions.  So I did, and updating went fine.
4. But now I decided to download Valhalla data for these regions as well.  So I installed Valhalla from OpenRepos, activated it in OSM Scout Server's settings, but "Check for updates" does not cause anything new to be downloaded.  Unsubscribing and re-subscribing all regions, then checking for updates again does not help either.

**Comment by rinigus** [comment]
Have you tried whether Valhalla works? Since you might have downloaded the maps earlier. You could also check if there are files under your maps folder in `valhalla/packages` and `valhalla/tiles`

**Comment by Olf0** [comment]
> Have you tried whether Valhalla works?

Well, OSM Scout Server's log states, Valhalla is started, it later complains about the missing elevation data, but nothing else.  So I assume Valhalla is running properly.   Check with `ps aux | grep valhalla`:
`nemo     28307  0.3  4.2 124712 35492 ?        Sl   20:19   0:00 /usr/bin/harbour-osmscout-server-module-route /home/nemo/.cache/harbour-osmscout-server/harbour-osmscout-server/valhalla/valhalla.json 1`
Without any data for Valhalla, routing must fail, if triggered from Poor / WhoGo Maps (haven't tried yet).

> Since you might have downloaded the maps earlier. You could also check if there are files under your maps folder in valhalla/packages and valhalla/tiles

The folder *valhalla* does not exist, as nothing was ever downloaded.

**Comment by rinigus** [comment]
OK, since you are on Custom, then:

* in Map Manager, enable "Store datasets for Valhalla..." . This should immediately show you that the data is missing.

* After downloading, go to Settings and choose Valhalla as a Routing engine.

This is expected to work. Please let me know how it went

**Comment by Olf0** [comment]
And again just my own lack of oversight: "Use Valhalla" (in the settings) was on, but "Store datasets for Valhalla" (in the map manager) not.

And I discovered a subdirectory *osmscout/europe-germany* (i.e. not *osmscout/europe-germany-\**), which seems to be stale (last touched 2018-02-19).  All other subdirectories there were updated today, plus *valhalla/\** just now.

**Comment by rinigus** [comment]
@Olf0, yes, the directories are kept even after cleanup and if they stay empty. I wasn't very comfortable writing that part since it may lead to the loss of data => I was doing my best to be cautious.

**Comment by Olf0** [comment]
@rinigus, first of all:
> [...] I was doing my best to be cautious.

... and you succeeded!  As

**Comment by rinigus** [comment]
Thank you! :) 

There might be some old mapbox gl packages as well and it maybe deleting them. Other than that, maybe packages were not registered properly from yesterday downloads. Try to cleanup and let's see what will happen.

**Comment by Olf0** [comment]
By timestamps (modification time) all (sqlite) files in *mapboxgl/packages* are from yesterday.

But ai solved another self posed conundrum: *osmscout/europe-germany* was downloaded with OSM Scout, which shares (via a symbolic link) the *osmscout* subdirectory with OSM Scout Server on my SD-card.  I remember doing this on purpose, as this way I was able to view and route within whole Germany (or even D-A-CH with the appropriate map downloaded in OSM Scout) using Poor Maps and OSM Scout Server (with the small profile "libosmscout with Geocoder-NLP").
But there is also a *geocoder-nlp/europe-germany* subdirectory with the same timestamp (2018-02-19), which should be out of reach for OSM Scout (?!?).  Does OSM Scout Server's Map Manager complement existing *osmscout/\** directories with corresponding *geocoder-nlp/\** data?  If so, really cool.  If not, another conundrum (I potentially tricked myself into).

**Comment by Olf0** [comment]
Executing the cleanup did:

* Delete *osmscout/europe-germany* (as expected).
* Empty *geocoder-nlp/europe-germany* (and setting its timestamp to now).
* Likely delete a few files in *mapboxgl/packages* (as announced).

> *osmscout/europe-germany* was downloaded with OSM Scout [...] to view and route within whole Germany (or even D-A-CH with the appropriate map downloaded in OSM Scout) using Poor Maps and OSM Scout Server (with the small profile "libosmscout with Geocoder-NLP").

At least this was what I was trying to achieve.  As OSM Scout Server did not let me select *osmscout/europe-germany* right before I deleted it an hour ago (probably because it is missing in *countries_requested.json*), I guess I forgot about it in middle of experimenting.

Anyway, OSM Scout was always working fine (and still is) with libosmscout map data managed by OSM Scout Server.   IIRC, I once (about a year ago) tried to remember the rule "Solely use OSM Scout Server' Map Manager for managing libosmscout maps shared with OSM Scout" and ultimately forgot about it.

**Comment by Olf0** [comment]
Two questions (one I wanted to pose for long), slightly off topic, but still related to OSM Scout Server:

1. I there a technical reason why OSM Scout Server cannot use multiple libosmscout maps simultaneously, in contrast to OSM Scout (with the same maps)?
   It is really annoying that currently one has to select a specific map in OSM Scout Server to be displayed in an OSM Scout Server client app.
2. Wouldn't distributing the glyphs for Mapnik maps (Noto font) the same way as the glyphs for MapboxGL maps achieve the same benefits as (as currently) a separate RPM packaged "module", plus additional ones?
    * Separately upgradeable from OSM Scout Server's RPM.
    * But not necessarily stored on internal FLASH memory (eMMC), instead in a directory of users' choice.
    * AFAIU a separate *glyphs* directory also allows for updating them independent of the maps proper.

But as usual in the last couple of days, I likely missed something contradictory (which I probably ought to know / once knew), again.

**Comment by rinigus** [comment]
@Olf0 - thank you for detailed reports! It seems that all is working fine now.

Re general backend support: It looks to me that right now the best backends are Mapbox GL/Valhalla/Geocoder NLP. Lot has changed during the year-and-a-half of the development, with me switching from libosmscout to Mapnik for better rendering, GeocoderNLP for better search, and, finally, moving to MapboxGL for fast and high quality rendering. Since I am competing with the big development houses doing maps on Android, I have to be practical and focus my development on what I think is the best and where it seems that we are behind.

Re libosmscout support: Every backend requires time. APIs change, new features come. The last update of libosmscout was done in the beginning of summer 2017 which broke maps imports badly. This was during preparation of 1.0 release. So, I had to revert back to May 2017 and kept the library version there. As a result, new features, such as better rendering of multiple maps, routing between 2 maps next to each other, are not supported. I asked whether there is someone who wants to maintain libosmscout backend in OSM Scout Server (https://sourceforge.net/p/libosmscout/mailman/message/36160373/), but, while there was a discussion about it, we don't have a maintainer at this moment of time.

For the best libosmscout experience, I would recommend OSM Scout by Karry - I am sure he will be able to provide full navigation support one day as well.

Re glyphs for Mapnik: Then see my reply too. In short, it looks that this will be waste of time in long term since it will be resolved automatically by people not using these fonts and uninstalling them. Hence, its better to focus on things that have future - there are lots of issues to be addressed with search, for example.

**Comment by Olf0** [comment]
@rinigus, thanks again for your kind and comprehensive support, both specifically to the original issue and in general at TMO and here.

> Re general backend support: It looks to me that right now the best backends are Mapbox GL/Valhalla/Geocoder NLP.

"The best" depends on use cases and technical restrictions (CPU, RAM and especially storage space).
In terms of look and feel (i.e. "user experience") of the base maps on a modern device with a large SD-card your statement is absolutely true.

* Although SailfishX seems to have gained some traction, the majority of SailfishOS users still own a Jolla 1 phone or Intex Aquafish / Inoi R7 / Jolla C, which are ultra-lowend Smartphone hardware by today's standards: 16 GB eMMC (much less free), 1 / 2 GB RAM, two midrange ("Krait", A15) / four low-end (A7) cores.
* While the Xperia X provides more RAM and processing powered, it does not provide much more free spaces (<< 16 GB, AFAIU).  The XA2 is not more capable and the big Accione "Jalla" phone at the same level.
* But in 2018 also the small Accione "Jalla" phone (with SFOS), the Inoi Tablets (7" and 10") and the rugged Inoi phone are going to be released, which basically provide the capabilities of Intex Aquafish / Inoi R7 / Jolla C.  Hence ultra-low end hardware IMO still should be well supported by OSM Scout Server.
* Also take device concepts as the SailfishOS Feature Phone shown at MWC 2018 and the Jolla Watch into account, which bear even weaker hardware, if ever put into production.
* SD-cards with "64 GB" or more are still costly for e.g. students or inhabitants of "developing countries" (e.g. BRICS countries).  Judging by TJC a lot of SailfishX users are quite young and the Intex Aquafish was / the Accione Jalla phones with SFOS will aim at those markets.
* Corporate and governmental users are often not allowed to insert a SD-card in order to prevent data "loss" (= leakage).
* The maps add up in size: Living in the Rhine-Ruhr area, I would like to have permanent offline data for at least NL, BE and whole DE, plus temporally varying travel locations.
A "32 GB" SD-card with some music on it is barely sufficient for storing these data sets, even when using the smallest, fully functional "profile": *libosmscout with Geocoder-NLP* (using Mapnik or MapboxGL data is infeasible).
* The same applies to a typical inhabitant of a BE-NE-LUX country (wants BE, NL, LU plus FR and / or DE permanently) or D-A-CH area (wants DE, CH and AT permanently).  Or imagine a Russian governmental (e.g. Russian Post) Inoi user, who wants to cover the whole Russian Federation plus some neighboring countries.
* If I consider covering e.g. Western Europe (at least DK, DE, AT, CH, NL, BE, LU, FR, IT) with Mapnik or MapboxGL data, plus data for Valhalla and libpostal, a "64 GB" SD-card solely used for OSM Scout Server data is barely sufficient.  Then think of a sales representative for a company, who has to cover the whole EU!
* The current vector map client (WhoGo Maps) [lacks overlay maps](https://github.com/otsaloma/whogo-maps/issues/31).  This loss of functionality IMO constitutes a regression.
 
> Lot has changed during the year-and-a-half of the development, with me switching from libosmscout to Mapnik for better rendering, GeocoderNLP for better search, and, finally, moving to MapboxGL for fast and high quality rendering. Since I am competing with the big development houses doing maps on Android, I have to be practical and focus my development on what I think is the best and where it seems that we are behind.

Absolutely agreed that priorities must be set, especially as resources are very limited.

Side note: Take a step back and consider what you already achieved within OSM Scout Server and its clients!  The SFOS native Here Maps is long left behind, the only remaining map software to look up to is OSMand~!  All others are not FLOSS, fully offline capable (many both) or lack other functionality.  Furthermore OSMand has become a monster in terms of APK size (~ 80 MB, although for multiple CPU architectures) and required processing power (still works fine on a Jolla 1 phone, although quite sluggish with recent releases, e.g. when scrolling and panning a already feature reduced map).  Plus there is no AlienDalvik on all corporate & governmental Inoi devices (deliberately) and  those with a community port of SFOS.

> Re libosmscout support: Every backend requires time. APIs change, new features come.

As every backend induces a maintenance burden (code, updates of external libraries, support requests, server space, download volume etc.), obviously its support in OSM Scout Server must be justified.
Hence considering to eliminate the Mapnik backend sounds well reasonable, as Mapnik seems to be inferior to MapboxGL (size, rendering quality, rendering speed).

But as novel client-server approach (locally) of OSM Scout Server allows for various map clients aiming at different use cases, IMO a (single) fully functional backend (or set of backends) requiring just a small dataset should be maintained in parallel to a "high-end" backend / or set of backends.
Currently the profile "libosmscout with Geocoder-NLP" nicely fulfills the first role, needing less than half the storage space than the "high-end" combination MapboxGL, Valhalla and libpostal / Geocoder-NLP.

Side note: Actually the "various map clients aiming at different use cases" idea pans out with WhoGo Maps, modRana, Poor Maps and Laufhelden. :) 

> The last update of libosmscout was done in the beginning of summer 2017 which broke maps imports badly. This was during preparation of 1.0 release. So, I had to revert back to May 2017 and kept the library version there.

Oh, I remember you mentioning this back then, but these issues should be settled by now, right?

> As a result, new features, such as better rendering of multiple maps, routing between 2 maps next to each other, are not supported. I asked whether there is someone who wants to maintain libosmscout backend in OSM Scout Server (https://sourceforge.net/p/libosmscout/mailman/message/36160373/), but, while there was a discussion about it, we don't have a maintainer at this moment of time.

I am not trying to suggest to invest much time into libosmscout right now, but not to discard it and at least maintain it on the current level of functionality (maps and routing), which is O.K. despite the lack of supporting adjacent maps.  Once ditched a reintegration will likely be even harder (maybe less for you than for an external contributor).

> For the best libosmscout experience, I would recommend OSM Scout by Karry - I am sure he will be able to provide full navigation support one day as well.

Now that I have bought a "128 GB" SD-card a few months ago and have tested the mapboxgl-qml library successfully on my Jolla 1 phone a few days ago, I will continue to pursue the "high-end" native solution to retire OSMand (currently I still fail to have TTS working stably in WhoGo & Poor Maps and WhoGo Maps lacks nautical etc. overlay maps).
BTW, on a Jolla 1 phone with its low (960x540) display resolution, Poor Maps, modRana, OSM Scout and WhoGo Maps basically provide the same user experience (speed and rendering quality), it is rather the small differences (e.g. TTS, client usability, which overlay maps are  available etc.) which have become of importance, now that all cover the basic functionality well.  BTW, navigation works in OSM Scout,
but no TTS (and have not checked if completely offline capable).

> Re glyphs for Mapnik: Then see my reply too. In short, it looks that this will be waste of time in long term since it will be resolved automatically by people not using these fonts and uninstalling them.

Now that  you have adjusted the desription of the Fonts Module on OpenRepos and display the "missing" warning only when Mapnik data is subscribed, this expectation is absolutely reasonable.

> Hence, its better to focus on things that have future - there are lots of issues to be addressed with search, for example.

Yes, but there are short term ("tactical") and long term ("strategical") aspects as well as technical, organisational (maintenance, distribution logistics), usability for different use cases and marketing ones.
To provide users with more freedom when inputting search terms ("fuzzy search") is sure one of the most pressing tactical, technical & usability goals, my intention here is to point to potential other goals (which do not necessarily conflict with each other).

**Comment by M4rtinK** [comment]
Just some notes from my position as Sailfish OS user and modRana developer.

> Although SailfishX seems to have gained some traction, the majority of SailfishOS users still own a Jolla 1 phone or Intex Aquafish / Inoi R7 / Jolla C, which are ultra-lowend Smartphone hardware by today's standards: 16 GB eMMC (much less free), 1 / 2 GB RAM, two midrange ("Krait", A15) / four low-end (A7) cores.

This would be actually pretty interesting to know - it seems to me many new and previously ex-Maemo/Sailfish faces have shown up/came back post Sailfish X. Also some of my friends finally started using Sailfish with Sailfish X as they have been previously unconvinced by the relatively lackluster hardware & it's bad availability.

Unfortunately I don't know about any reasonable data sources for this - Jolla never every publishes any absolute sale or active installation numbers a statistics in Harbour and in OpenRepos only show accumulated download numbers per app, nothing about the originating device.

> SD-cards with "64 GB" or more are still costly for e.g. students or inhabitants of "developing countries" (e.g. BRICS countries). Judging by TJC a lot of SailfishX users are quite young and the Intex Aquafish was / the Accione Jalla phones with SFOS will be aimed at those markets.

I'm not sure this is really a valid point - looking at Amazon:
https://www.amazon.com/Micro-SD-Memory-Cards/b?ie=UTF8&node=3015433011
Looks like 64 GB micro SD cards (I still have a 16 MB SD card somewhere...) cost about 20-30$, which does not seem that much as you have just shelled out either 300$ + 50€ for Sailfish X or 140-300$ for an Accione. Just does not seem that much for something as critical as reasonable amount of offline storage & backup space (in case of phone brick), not to mention for storing all the highress photos/videos modern phones can take.

> Corporate and governmental users are often not allowed to insert a SD-card in order to prevent data "loss" (= leakage).

Yep, that's definitely a better argument and indeed a valid policy in security sensitive areas.

> The current vector map client (WhoGo Maps) lacks overlay maps. This loss of functionality IMO constitutes a regression.

For the record I consider map overlays critical functionality that modRana has supported since almost its inception back in 2010 and I definitely plan to include map overlay support wile adding support for using the MapBox GL map widget.

Actually the plan is to first add MapBox GL as an alternative map widget people can switch to while the current PinchMap based tile widget is the default. Once MapBox GL reaches feature parity from the modRana PoV, it will become the default map widget with the current one acting as backup & used on platforms not supported by MapBox GL (if any).

**Comment by rinigus** [comment]
This discussion is becoming rather general and maybe of interest to others as well. I suggest to move future posts to TMO.

As for libosmscout - I have made it possible to compile the server without its support. Same goes to most of the other backends and would allow users to select the backends on their Linux PCs if needed. For SFOS, I plan to keep libosmscout as it is, without any new functionality unless someone will step up and start maintaining it. Mapnik is very mature and requires almost no maintenance. I am not going to work on its styles though, to keep it at low maintenance mode. Since Mapnik is way ahead of libosmscout rendering, it will be strange to drop it. Maybe someone would prefer using it on more powerful devices/PCs.

So, to summarize, there are no plans to remove backends from the code.
